### PR TITLE
Add translation of Truffle arguments to symbolic names for TruffleRuby.

### PR DESCRIPTION
**Pull request by nirvdrum**
This PR hooks into the existing Truffle pass to provide non-indexed names for Truffle languages that pack their arguments in a consistent format. I've only added a TruffleRuby implementation in this pass, but the machinery is there for other languages.

The graphs are fairly large, so I'm using screenshots rather than mermaid here. All of the translated `LoadIndexed` nodes that normally get turned into `T(<int>)` now get turned into `T(<packed_arg_name>)`. If a language doesn't have a translator implementation, the previous `T(<int>)` format is used.

The mechanism to detect the language is a fuzzy match because the graph data doesn't directly indicate the language. I've used a breadth-first search following _out_ edges from the `LoadIndexed` node. If there's a more reliable search algorithm, I'm happy to adjust. I'm also not heavily invested in any of the names here, so if "translator" is too broad, please suggest an alternative and I'll update the PR.

### Before

image

### After

image

### CODE REVIEW

File: lib/seafoam/passes/truffle_translators/default.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+require "set"
```
Comment (chrisseaton) [inline lib/seafoam/passes/truffle_translators/default.rb:3]: Not used?

File: lib/seafoam/passes/truffle_translators/translators.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      TRUFFLE_LANGUAGES = {
+        "org.truffleruby" => "TruffleRuby",
+      }
+
+      class << self
+        def get_translator(node)
+          queue = node.outputs.collect(&:to)
+          visited = Set.new
+
+          until queue.empty?
```
Comment (chrisseaton) [inline lib/seafoam/passes/truffle_translators/translators.rb:20]: Can we abstract this into a method which yields nodes, and one which looks at them?

File: lib/seafoam/passes/truffle_translators/truffleruby.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+module Seafoam
+  module Passes
+    module TruffleTranslators
+      class TruffleRuby < Default
+        TRUFFLERUBY_ARGS = ["DECLARATION_FRAME", "CALLER_FRAME_OR_VARIABLES", "METHOD", "DECLARATION_CONTEXT",
+                            "FRAME_ON_STACK_MARKER", "SELF", "BLOCK", "DESCRIPTOR",]
+
+        def translate_argument_load(index)
+          index >= TRUFFLERUBY_ARGS.size ? "Arg[#{index - TRUFFLERUBY_ARGS.size}]" : TRUFFLERUBY_ARGS[index]
```
Comment (chrisseaton) [inline lib/seafoam/passes/truffle_translators/truffleruby.rb:11]: Minor but I think I'd stylistically prefer `args[n]`.

File: lib/seafoam/passes/truffle.rb [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 require "tsort"
 
+require_relative "truffle_translators/translators"
```
[the remark this answers is not in the document]
Comment (nirvdrum) [reply lib/seafoam/passes/truffle.rb:5]: `require` searches the `$LOAD_PATH` while `require_relative` searches relative to the current file. Since I'm not trying to load from an external library, `require_relative` seemed most appropriate.

File: lib/seafoam/passes/truffle_translators/default.rb (hunk 1 above)
Comment (nirvdrum) [reply lib/seafoam/passes/truffle_translators/default.rb:3]: Bad refactoring. It should be in the _translators.rb_ file. I'll delete it.

**Comment by eregon** [comment]
Very cool :)

File: lib/seafoam/passes/truffle_translators/translators.rb (hunk 2 above)
Comment (nirvdrum) [reply lib/seafoam/passes/truffle_translators/translators.rb]: I've refactored a bit and extracted a BFS search search strategy to accompany a new `visit` and `visit_outputs` method pair in 8c535fe7724a58bc370647fefed47b599ed3cd8f. @eregon Do you think this satisfies the design change request?

File: lib/seafoam/graph.rb [hunk 5]
```diff
@@ -87,6 +89,18 @@ def id_and_label
     def inspect
       "<Node #{id} #{node_class}>"
     end
+
+    def visit(&block)
+      block.call(self)
+    end
```
Comment (eregon) [inline lib/seafoam/graph.rb:95]: This could be inlined, but it also adds some value for readability so seems fine.

File: lib/seafoam/passes/truffle_translators/translators.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        "org.truffleruby" => "TruffleRuby",
+      }
+
+      class << self
+        def get_translator(node)
+          translator = node.visit_outputs(:bfs) do |entry|
+            declaring_class = (entry.props["code"] || entry.props.dig(
+              "nodeSourcePosition",
+              :method,
+            ))&.[](:declaring_class)
```
Comment (eregon) [inline lib/seafoam/passes/truffle_translators/translators.rb:21]: This looks quite cryptic and is hard to parse.
How about:
```suggestion
            declaring_class = entry.props["code"] || entry.props.dig("nodeSourcePosition", :method)
            declaring_class = declaring_class[:declaring_class] if declaring_class
```

File: lib/seafoam/passes/truffle_translators/truffleruby.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module Seafoam
+  module Passes
+    module TruffleTranslators
+      class TruffleRuby < Default
+        TRUFFLERUBY_ARGS = [
+          "DECLARATION_FRAME",
+          "CALLER_FRAME_OR_VARIABLES",
```
Comment (eregon) [inline lib/seafoam/passes/truffle_translators/truffleruby.rb:9]: This never contains the the caller frame anymore, but we didn't rename it yet in RubyArguments and simplify it.
It should be:
```suggestion
          "CALLER_SPECIAL_VARIABLES",
```

**Review by eregon — APPROVED** [review]

File: lib/seafoam/graph.rb (hunk 5 above)
Comment (nirvdrum) [reply lib/seafoam/graph.rb:95]: Right. I only added it to use somewhat standard naming for the visitor pattern.

File: lib/seafoam/passes/truffle_translators/truffleruby.rb (hunk 7 above)
Comment (nirvdrum) [reply lib/seafoam/passes/truffle_translators/truffleruby.rb]: Thanks for the correction.
